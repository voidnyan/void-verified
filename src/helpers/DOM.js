export class DOM {
	static create(element, classes = null, children = null) {
		const el = document.createElement(element);
		if (classes !== null) {
			for (const className of classes?.split(" ")) {
				if (className.startsWith("#")) {
					el.setAttribute("id", `void-${className.slice(1)}`);
					continue;
				}
				el.classList.add(`void-${className}`);
			}
		}

		if (children) {
			try {
				el.append(...children);
			} catch {
				el.append(children);
			}
		}

		return el;
	}

	static getOrCreate(element, classes) {
		const id = classes
			.split(" ")
			.find((className) => className.startsWith("#"));
		return this.get(id) ?? this.create(element, classes);
	}

	static get(selector) {
		return document.querySelector(selector);
	}

	static getAll(selector) {
		return document.querySelectorAll(selector);
	}
}
