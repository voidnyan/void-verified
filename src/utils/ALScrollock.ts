export class ALScrollock {
	static lock() {
		document.body.classList.add("scroll-lock");
	}

	static unlock(){
		document.body.classList.remove("scroll-lock");
	}
}
