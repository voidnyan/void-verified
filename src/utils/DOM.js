export class DOM {
	static create(element, classes = null, children = null, options = {}) {
		const el = document.createElement(element);
		if (classes !== null) {
			for (const className of classes?.split(" ")) {
				if (className.startsWith("#")) {
					el.setAttribute("id", `void-${className.slice(1)}`);
					continue;
				}
				el.classList.add(`void-${className}`);
			}
		}

		if (children) {
			if (Array.isArray(children)) {
				el.append(...children);
			} else {
				el.append(children);
			}
		}

		for (const key of Object.keys(options)) {
			el[key] = options[key];
		}

		return el;
	}

	static render(element, parent) {
		const htmlElement = document.createElement(element);
		parent.append(htmlElement);
	}

	static renderReplace(element, parent) {
		const htmlElement = document.createElement(element);
		htmlElement.setAttribute("title", "wow");
		parent.replaceChildren(htmlElement);
	}

	static getOrCreate(element, classes) {
		const id = classes
			.split(" ")
			.find((className) => className.startsWith("#"));
		return this.get(id) ?? this.create(element, classes);
	}

	static get(selector) {
		const convertedSelector = this.#convertSelector(selector);
		return document.querySelector(convertedSelector);
	}

	static getAll(selector) {
		const convertedSelector = this.#convertSelector(selector);
		return document.querySelectorAll(convertedSelector);
	}

	static #convertSelector(selector) {
		let results = [];
		for (const className of selector?.split(" ")) {
			if (className.startsWith("#")) {
				results.push(`#void-${className.slice(1)}`);
				continue;
			}
			results.push(`.void-${className}`);
		}
		return results.join(" ");
	}
}
