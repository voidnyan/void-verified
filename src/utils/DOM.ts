import {Vue} from "./vue";

export class DOM {
	static create<T extends HTMLElement = HTMLElement>(element: string, classes?: string, children?: any, options = {}) : T {
		const el = document.createElement(element) as T;

		if (element.toLowerCase() === "a") {
			el.addEventListener("click", (event) => {
				Vue.handleAnchorClickEvent(event);
			})
		}

		this.transformClasses(el, classes);

		if (children) {
			if (Array.isArray(children)) {
				// @ts-ignore
				el.append(...children);
			} else {
				// @ts-ignore
				el.append(children);
			}
		}

		for (const key of Object.keys(options)) {
			el[key] = options[key];
		}

		return el;
	}

	static createAnchor(href, classes = null, children = null): HTMLAnchorElement {
		const anchor = this.create<HTMLAnchorElement>("a", classes, children);
		anchor.setAttribute("href", href);
		return anchor;
	}

	static createDiv(classes?: string, children?: any): HTMLDivElement {
		return DOM.create<HTMLDivElement>("div", classes, children);
	}

	static render(element, parent) {
		const htmlElement = document.createElement(element);
		parent.append(htmlElement);
	}

	static renderReplace(element, parent) {
		const htmlElement = document.createElement(element);
		htmlElement.setAttribute("title", "wow");
		parent.replaceChildren(htmlElement);
	}

	static getOrCreate(element, classes) {
		const id = classes
			.split(" ")
			.find((className) => className.startsWith("#"));
		return this.get(id) ?? this.create(element, classes);
	}

	static get(selector) {
		const convertedSelector = this.#convertSelector(selector);
		return document.querySelector(convertedSelector);
	}

	static getAll(selector) {
		const convertedSelector = this.#convertSelector(selector);
		return document.querySelectorAll(convertedSelector);
	}

	static transformClasses(element: HTMLElement, classes?: string) {
		if (classes) {
			for (const className of classes?.split(" ")) {
				if (className.startsWith("#")) {
					element.setAttribute("id", `void-${className.slice(1)}`);
					continue;
				}
				if (className.startsWith(".")) {
					element.classList.add(className.slice(1));
					continue;
				}
				element.classList.add(`void-${className}`);
			}
		}
	}

	static #convertSelector(selector) {
		let results = [];
		for (const className of selector?.split(" ")) {
			if (className.startsWith("#")) {
				results.push(`#void-${className.slice(1)}`);
				continue;
			}
			results.push(`.void-${className}`);
		}
		return results.join(" ");
	}
}
