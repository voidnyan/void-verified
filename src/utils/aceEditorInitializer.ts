export class AceEditorInitializer {
	static initializeEditor(id: string, value: string) {
		const siteTheme = localStorage.getItem("site-theme");
		let theme = "monokai";
		switch (siteTheme) {
			case "dark":
			case "system":
				theme = "monokai";
				break;
			case "contrast":
			default:
				theme = "dawn";
		}
		// @ts-ignore
		const editor = ace.edit(id, {
			mode: "ace/mode/css",
			theme: `ace/theme/${theme}`,
			value: value
		});

		editor.setKeyboardHandler("ace/keyboard/vscode");

		editor.setOptions({
			useWorker: false,
			enableBasicAutocompletion: true,
			highlightSelectedWord: true,
			copyWithEmptySelection: true,
			scrollPastEnd: true,
			showPrintMargin: false
		});
	}

	static addChangeHandler(id: string, callback: (value: string) => void) {
		// @ts-ignore
		const editor = ace.edit(id)
		editor.on("change", () => {
			callback(editor.getValue());
		})
	}
}
