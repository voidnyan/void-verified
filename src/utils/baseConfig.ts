export class BaseConfig {
	configInLocalStorage: string;
	constructor(configInLocalStorage: string) {
		this.configInLocalStorage = configInLocalStorage;
	}
	save() {
		localStorage.setItem(this.configInLocalStorage, JSON.stringify(this));
	}

}
