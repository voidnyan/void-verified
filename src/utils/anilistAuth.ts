import {imageHosts, ImageHostService} from "../api/imageHostConfiguration";
import {ImgurAPI} from "../api/imgurAPI";
import {DOM} from "./DOM";
import {StaticSettings} from "./staticSettings";
import {ButtonComponent} from "../components/ButtonComponent";
import {VoidApi} from "../api/voidApi";

export class AnilistAuth {
	private static localStorageAuth = "void-verified-auth";
	static token: string = null;
	static expires: Date = null;

	static name: string;
	static id: number;

	private static settingsContainer = DOM.createDiv();

	static initialize() {
		const auth = JSON.parse(localStorage.getItem(this.localStorageAuth)) ?? null;
		if (auth) {
			this.token = auth.token;
			this.expires = new Date(auth.expires);
		}


		const alAuth = JSON.parse(localStorage.getItem("auth"));
		this.name = alAuth?.name;
		this.id = alAuth?.id;
	}

	static checkAuthFromUrl() {
		const hash = window.location.hash.substring(1);
		if (!hash) {
			return;
		}

		const [path, token, type, expiress] = hash.split("&");

		if (path === "void_imgur") {
			const imgurConfig =
				ImageHostService.getImageHostConfiguration(
					imageHosts.imgur,
				);
			new ImgurAPI(imgurConfig).handleAuth();
		}

		if (path === "void_api_auth") {
			VoidApi.token = token;
			localStorage.setItem("void-verified-api-token", token);
		}

		if (path === "void_auth") {
			const expiresDate = new Date(
				new Date().getTime() + Number(expiress.split("=")[1]) * 1000,
			);

			this.saveAuthToken({
				token: token.split("=")[1],
				expires: expiresDate,
			});
		}

		window.history.replaceState(
			null,
			"",
			"https://anilist.co/settings/developer",
		);
	}

	static saveAuthToken(tokenObject: {token: string, expires: Date}) {
		this.token = tokenObject.token;
		this.expires = tokenObject.expires;
		localStorage.setItem(
			this.localStorageAuth,
			JSON.stringify(tokenObject),
		);
	}

	static createSettings() {
		this.renderSettings();
		return this.settingsContainer;
	}

	private static renderSettings() {
		this.settingsContainer.replaceChildren();
		const isAuthenticated =
			this.token !== null &&
			new Date(this.expires) > new Date();

		const clientId = 15519;

		const header = DOM.create("h3", null, "Authorize (AniList API)");
		const description = DOM.create(
			"p",
			null,
			"Some features of VoidVerified might need your access token to work correctly or fully. Below is a list of features using your access token. If you do not wish to use any of these features, you do not need to authorize. If revoking authentication, be sure to revoke VoidVerified from Anilist Apps as well.",
		);

		const list = DOM.create("ul");
		for (const option of Object.values(StaticSettings.options).filter(
			(o) => o.authRequired,
		)) {
			list.append(DOM.create("li", null, option.description));
		}

		// DOM.create uses Vue router so don't use that here
		const authLink = document.createElement("a");
		authLink.classList.add("void-button");
		authLink.append("Authorize VoidVerified");
		authLink.setAttribute(
			"href",
			`https://anilist.co/api/v2/oauth/authorize?client_id=${clientId}&response_type=token`,
		);

		const removeAuthButton = DOM.create(
			"button",
			null,
			"Revoke auth token",
		);
		removeAuthButton.classList.add("button");
		removeAuthButton.addEventListener("click", () => {
			this.removeAuthToken();
			this.renderSettings();
		});

		this.settingsContainer.append(header);
		this.settingsContainer.append(description);
		this.settingsContainer.append(list);
		this.settingsContainer.append(
			!isAuthenticated ? authLink : removeAuthButton,
		);


		this.settingsContainer.append(VoidApi.createSettings());
	}

	static 	removeAuthToken() {
		this.token = null;
		this.expires = null;
		localStorage.removeItem(this.localStorageAuth);
	}
}
