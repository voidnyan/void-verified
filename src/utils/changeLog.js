import { Checkbox, Modal, Note, SettingLabel } from "../components/components";
import { changeLog } from "../assets/changeLog";
import { DOM } from "./DOM";

export class ChangeLog {
	#lastVersion;
	#settings;
	#lastVersionInLocalStorage = "void-verified-changelog-last-version";

	constructor(settings) {
		this.#settings = settings;
		this.#lastVersion = localStorage.getItem(
			this.#lastVersionInLocalStorage,
		);
	}

	renderChangeLog(forceDisplay = false) {
		if (
			!this.#settings.options.changeLogEnabled.getValue() &&
			!forceDisplay
		) {
			return;
		}

		if (!this.#newVersionExists() && !forceDisplay) {
			return;
		}

		const modalBody = [
			DOM.create(
				"div",
				"change-log-note",
				Note(
					"Here are some changes included in recent releases. You can enable new features here or later from settings. You can view this popup again or disable it from settings.",
				),
			),
		];
		modalBody.push(
			...changeLog.map((version) => {
				return this.#createModalContent(version);
			}),
		);

		document.body.append(
			Modal(modalBody, () => {
				this.#handleClose(this);
			}),
		);
	}

	#newVersionExists() {
		if (!this.#lastVersion) {
			return true;
		}
		const versions = changeLog.map((version) =>
			version.versionNumber.split("."),
		);
		const [lastMajorVersion, lastMinorVersion] =
			this.#lastVersion.split(".");

		for (const [majorVersion, minorVersion] of versions) {
			if (
				Number(majorVersion) > Number(lastMajorVersion) ||
				Number(minorVersion) > Number(lastMinorVersion)
			) {
				return true;
			}
		}

		return false;
	}

	#createModalContent(version) {
		const container = DOM.create("div");
		const header = DOM.create(
			"h3",
			"change-log-header",
			`Version ${version.versionNumber}`,
		);
		container.append(header);
		const list = DOM.create("ul", "change-log-list");
		const listItems = version.featureList.map((feature) => {
			return this.#createFeatureListItem(feature);
		});
		list.append(...listItems);
		container.append(list);
		return container;
	}

	#createFeatureListItem(feature) {
		const container = DOM.create("li");
		if (feature.option) {
			const value = this.#settings.options[feature.option].getValue();
			container.append(
				SettingLabel(
					feature.description,
					Checkbox(value, (event) => {
						this.#handleOptionChange(event, feature.option);
					}),
				),
			);
			return container;
		}
		container.append(
			DOM.create("span", "change-log-list-item", [
				DOM.create("span", null, "-"),
				DOM.create("span", null, feature.description),
			]),
		);
		return container;
	}

	#handleOptionChange(event, option) {
		const value = event.target.checked;
		this.#settings.saveSettingToLocalStorage(option, value);
	}

	#handleClose(_changeLog) {
		const version = changeLog[0].versionNumber;
		_changeLog.#lastVersion = version;
		localStorage.setItem(_changeLog.#lastVersionInLocalStorage, version);
	}
}
