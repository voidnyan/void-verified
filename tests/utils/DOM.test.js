import { DOM } from "../../src/utils/DOM";

describe("create", () => {
	it("creates an element", () => {
		const elem = DOM.create("div");
		expect(elem).toMatchSnapshot();
	});

	it("creates an element with an id", () => {
		const elem = DOM.create("div", "#test");
		expect(elem).toMatchSnapshot();
	});

	it("creates an element with a class", () => {
		const elem = DOM.create("div", "test");
		expect(elem).toMatchSnapshot();
	});

	it("creates an element with an id and multiple classes", () => {
		const elem = DOM.create("div", "#test test test2");
		expect(elem).toMatchSnapshot();
	});

	it("creates an element with a child element", () => {
		const childElement = DOM.create("div", "child");
		const elem = DOM.create("div", "parent", childElement);
		expect(elem).toMatchSnapshot();
	});

	it("creates an element with an array of children", () => {
		const children = [1, 2, 3, 4, 5].map((val) => {
			return DOM.create("div", "child", val);
		});
		const elem = DOM.create("div", "parent", children);
		expect(elem).toMatchSnapshot();
	});
});

describe("Getters", () => {
	it("getOrCreate creates a new element", () => {
		const element = DOM.getOrCreate("div", "#test");
		expect(element).toMatchSnapshot();
	});

	it("getOrCreate gets existing element", () => {
		const existingElement = DOM.create(
			"div",
			"#test",
			"this exists in the body",
		);
		document.body.append(existingElement);
		const element = DOM.getOrCreate("div", "#test");
		expect(element).toMatchSnapshot();
	});

	it("getAll gets all elements", () => {
		const existingElems = [1, 2, 3, 4, 5].map((val) => {
			return DOM.create("div", "test", val);
		});
		document.body.append(...existingElems);
		const elems = DOM.getAll("test");
		expect(elems).toMatchSnapshot();
	});
});
