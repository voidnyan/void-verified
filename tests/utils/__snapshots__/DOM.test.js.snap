// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Getters getAll gets all elements 1`] = `
NodeList [
  <div
    class="void-test"
  >
    1
  </div>,
  <div
    class="void-test"
  >
    2
  </div>,
  <div
    class="void-test"
  >
    3
  </div>,
  <div
    class="void-test"
  >
    4
  </div>,
  <div
    class="void-test"
  >
    5
  </div>,
]
`;

exports[`Getters getOrCreate creates a new element 1`] = `
<div
  id="void-test"
/>
`;

exports[`Getters getOrCreate gets existing element 1`] = `
<div
  id="void-test"
>
  this exists in the body
</div>
`;

exports[`create creates an element 1`] = `<div />`;

exports[`create creates an element with a child element 1`] = `
<div
  class="void-parent"
>
  <div
    class="void-child"
  />
</div>
`;

exports[`create creates an element with a class 1`] = `
<div
  class="void-test"
/>
`;

exports[`create creates an element with an array of children 1`] = `
<div
  class="void-parent"
>
  <div
    class="void-child"
  >
    1
  </div>
  <div
    class="void-child"
  >
    2
  </div>
  <div
    class="void-child"
  >
    3
  </div>
  <div
    class="void-child"
  >
    4
  </div>
  <div
    class="void-child"
  >
    5
  </div>
</div>
`;

exports[`create creates an element with an id 1`] = `
<div
  id="void-test"
/>
`;

exports[`create creates an element with an id and multiple classes 1`] = `
<div
  class="void-test void-test2"
  id="void-test"
/>
`;
